File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.py
class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        nums.sort()
        left, right, ans = 1, nums[len(nums) - 1], 0
        
        while left <= right:
            mid = (left + right) // 2
            total = 0
            for num in nums:
                total += math.ceil(num/mid)
            if total <= threshold:
                ans, right = mid, mid - 1
            else:
                left = mid + 1
        return ans

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
    int smallestDivisor(vector<int>& nums, int threshold) {
        
        auto f = [&](int x) {
            int cur = 0;
            for (int i = 0; i < (int) nums.size(); i++) {
                cur += (nums[i] - 1) / x + 1;
            }
            
            return cur <= threshold;
        };
        
        int lo = 1, hi = *max_element(nums.begin(), nums.end());
        
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (f(mid))
                hi = mid;
            else
                lo = mid + 1;
        }
        return lo;
    }
};